### ISSUE TITLE
Blocking bucket creation

### ISSUE THREAD

**Issue by sarathbabur** [vladimir-bukhtoyarov/bucket4j#80]
Hi,

I am creating Bucket in Single Hazelcast instance serving two Application data centers (DC) and having custom waiting code to consumer the token as follows.

### Creation:
```java
Bandwidth limit = Bandwidth.classic(bucketSize, Refill.intervally(bucketSize, period));
Bucket bucket = Bucket4j.extension(Hazelcast.class).builder().addLimit(limit)
					.build(hazelCastService.getBucketMap(), bucketId, RecoveryStrategy.RECONSTRUCT);
```

### Consumption:
```java
while (true) {
	probe = getBucket().tryConsumeAndReturnRemaining(bucketBean.getTokenSize());
	if (probe.isConsumed()) {
		break;
	} else {
		Thread.sleep((long) (probe.getNanosToWaitForRefill() / 1000000.0));
	}
}
```

But the issue is first data centers consumes token faster like three times higher than other one. So there is no equal distribution of tokens.

When application threads from both DCs sleeps till the refill and wakes-up,  threads from first DC consumes faster than other one.

I am thinking it might be because custom handling of sleep and consume. 
Is this can be solved using "bucket.tryConsume(1, MAX_WAIT_NANOS, BlockingStrategy.PARKING)", so that I dont need to handle sleep. And token might being served in the order when thread arrives.

Please suggest if this works..

When it is, then I am not finding method from Bucket object, its part of BlockingBucket object, but I cannot able to instantiate that object. please help.

**Comment by vladimir-bukhtoyarov** [comment]
If I understand correctly, you want to park the thread until tokens will be available, if so then avoid the manual parking/sleeping, it will never work fairy in concurrent environment. Instead you should use functionality which buil-it to library:

```java
bucket.asScheduler().consume(tokens)
```
or
```java
boolean consumed = bucket.asScheduler().tryConsume(tokens, maximumDelayWhichYuaAreAgreeToWait)
```

The difference between your code and built-in functionality is following:
- Your code does not reserve tokens, so even after long waiting the thread can release the situation when another concurrent thread already stole the tokens while thread was sleeping.
- Scheduler API of Bucket4j solves the fairness problem, by reserving the tokens in the bucket, so independently from concurrency level nobody can steal the tokens belonged to parked thread.

Also I little bit worring about the ```getBucket()``` method, do not you cache the reference to the bucket? If not, then it is wrong, because creation the bucket on each request doubles the request rate to the Hazelcast server, one request to check that bucket exists and on request to do consumption request. If you are unable to cache the reference to bucket(for example because you do not know the amount of buckets that would be used), then you should use the ProxyManager as there https://github.com/vladimir-bukhtoyarov/bucket4j/blob/master/doc-pages/jcache-usage.md#example-1---limiting-access-to-http-server-by-ip-address  ProxyManager solves the double request rate problem.

**Comment by nitenA** [comment]
I have one question regarding this scheduler API. Assuming scheduler API solves bucket allocation algorithm fairly but just for my understanding, does it solves this problem locally within server or maintains that state within Bucket which is persisted in GRiD cache? If it’s locally, will this also consider threads running concurrently across instances?

Let’s say if I have 20 threads across instances and only 5 TPS available within bucket configuration? How scheduler API will decides that which thread to be given bucket because threads are across instances?

**Comment by nitenA** [comment]
Sorry to say that interface doesn’t clearly explains how tokens are reserved globally to handle concurrency. I am more interested in knowing the storage of reserved token to solve concurrency is inside bucket or locally within APi. If inside bucket, I should be able to lookup the threads and reserved bucket within bucket configuration?

Sorry one more question is ScehdulerApI internally uses BlockingBucket to achieve fairness?

**Comment by vladimir-bukhtoyarov** [comment]
> Sorry to say that interface doesn’t clearly explains how tokens are reserved globally to handle concurrency. 

Reservation is just just decreasing the amount of tokens in the bucket, nothing more. For example imagine that you have a bucket with limit 10 tokens per 10 second
```java
Bucket bucket = Bucket4j.builder()
        addLimit(Bandwidth.simple(10, Duration.ofSeconds(10)))
        .build();
```
Having any time **T** and 2 tokens in the bucket. Then any thread  tries to consume 5 tokens with limit of waiting 2 seconds ```bucket.tryConsume(5,  _2_seconds)```, result of operation will be **false**, amount of tokens in the bucket will be not changed, because 2 second  is not enough to close the deficit of 3 tokens.

But if thread is agree to wait more, for example ```bucket.tryConsume(5,  _10_seconds)```, then operation will be successful and will lead to following effects:
* Amount of tokens in the bucket will be -3, just ```2 - 5```, this is what I call reservation.
* Request immediately returned from server, nobody waits on server side, but thread(in the client JVM) that called tryConsume will be blocked for 3 second(the time that required to refill 3 token).

So, implementing reservation as simple decreasing is very effective solution because:
* Intercommunication between client threads or JVMs(in case of distributed usage) is not neccessary. 
* Threads of server that actually stores of bucket state is never parked, server processes requests as fast as possible.

> If inside bucket, I should be able to lookup the threads and reserved bucket within bucket configuration?

I do not understand the question, could you rephrase it?

**Comment by nitenA** [comment]
Thanks for explanation. If we have 10 concurrent threads who all at same point of time try to consumer token and wait for same time like 5 sec, I think still there is a possibility of race condition until unless API interrupts threads who ever requested token first. Which I still don’t understand how’s possible across JVMs when click time is different and grid bucket has no way of knowing which thread requested token first.

I do understand now that bucket always has token availabity state be it positive or negative based on availability but giving token to client is client side functionality which is part of your API.

At any point do u see race condition between threads specifically in distributed systems when threads are more than TPS availability.

**Comment by vladimir-bukhtoyarov** [comment]
> If we have 10 concurrent threads who all at same point of time try to consume token and wait for same time like 5 sec, I think still there is a possibility of race condition

The concrete way to solve concurrency problem depends from particular back-end, for example for in-memory buckets CAS is used by default, or ReentrantLock can be choosen as option. For JBoss Infinispan the ReadWriteMap+SerializableFunction is used, for Haselcast the IMap+EntryProccessor is used, stored procedures in Lua will be used for Redis, stored procedures in PL/SQL+pessimistic locking will be used for PostgreSQL, EntryProcessor is used for generic JCache implementation, etc... 
To summarize, Bucket4j uses the functionality from each back-end:
* That allow to execute the query in isolation(mutual exclusion).
* That allow to service any throttling request in one network-roundrip without reading bucket state form server.

So, you should not worry about concurrency and efficiently, the back-ends that does not satisfy the two requirements above will be never chosen as point of integration with Bucket4j, for example Bucket4j will never provide integration for Apache Cassandra because it does not provide functionality for writing stored procedures.

> Which I still don’t understand how’s possible across JVMs when clock time is different and grid bucket has no way of knowing which thread requested token first.

The client clock is never used, client just has no access to bucket state, the state modified strongly on the server that maintain bucket. It can be confusing for enterprise developers who expect the typical request processing flow like:
- Begin transaction
- Acquire lock if neccesarry
- Read data
- Modify and write data
- Free lock  if neccesarry
- Commit transaction.
But bucket4j never acts in this style. Today, if you want to achieve sub-milisecond latency(it is obviously that all clients expects from throttling library to be as fast as network allow) you should send calculation to place where data stored, instead of fetching data to place when calculation happens. If I understand correctly you are using Hazelcast, so I highly recomend to read about [EntryProcessor](https://docs.hazelcast.org//docs/latest-development/manual/html/Distributed_Computing/Entry_Processor/index.html) functionality if you want to understand how to Bucket4j works in distributed environment.

**Comment by nitenA** [comment]
Yes I am using Hazelcast and will go through EntryProcessor. Thanks very much for you time on answering questions. Will get back to you. Have a good weekend.

**Comment by vladimir-bukhtoyarov** [comment]
@sarathbabur I answered your question there https://github.com/vladimir-bukhtoyarov/bucket4j/issues/80#issuecomment-421539778

Please close the issue if you have no additional questions.

**Comment by nitenA** [comment]
Can u please answer last question which I asked above before issue is closed.

**Comment by vladimir-bukhtoyarov** [comment]
@nitenA feel free to ask

**Comment by vladimir-bukhtoyarov** [comment]
I do not use Hazelcast in my regular work, because GridGain is choosen, but integration of Bucket4j with Hazelcast is well tested by standard [TCK](https://github.com/vladimir-bukhtoyarov/bucket4j/blob/master/bucket4j-core/src/test/java/io/github/bucket4j/AbstractDistributedBucketTest.java) that used for testing all distrubuted back-ends, the high-concurrent scenario is tested as well. Also JHipster uses Hazelcast with Bucket4j https://www.jhipster.tech/api-gateway/#rate_limiting and many other like [spring-cloud-zuul-ratelimit](https://github.com/marcosbarbero/spring-cloud-zuul-ratelimit), so I can make decision that integration of Bucket4j with Hazelcast has no obvious problems. Of course you should do testing by yourself and read [this checklist](https://github.com/vladimir-bukhtoyarov/bucket4j/blob/master/doc-pages/production-jcache-checklist.md).

**Comment by nitenA** [comment]
Thanks. I am good with answers and will open new issue if have more questions.
